include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -306,10 +306,105 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+/**
+ * Helper function for creating a geometric coarsening sequence: Replicate a
+ * parallel::distributed::Triangulation completely in a serial
+ * triangulation. This can potentially be very memory-consuming, so the
+ * triangulation is only filled on the first rank of a compute node, i.e.,
+ * those MPI ranks where `rank % n_processes_per_node` is zero.
+ */
+template<int dim, int spacedim>
+void
+replicate_distributed_triangulation_on_node(
+  parallel::distributed::Triangulation<dim, spacedim> const & distributed_tria,
+  MPI_Comm const &                                            mpi_comm,
+  unsigned int const                                          n_processes_per_node,
+  dealii::Triangulation<dim, spacedim> &                      serial_tria)
+{
+  unsigned int const my_rank                  = Utilities::MPI::this_mpi_process(mpi_comm);
+  bool const         is_first_process_on_node = (my_rank % n_processes_per_node) == 0;
+
+  // copy level 0 of distributed triangulation
+  if(is_first_process_on_node)
+  {
+    auto [points, cell_data, sub_cell_data] =
+      GridTools::get_coarse_mesh_description(distributed_tria);
+
+    std::vector<std::pair<unsigned int, CellData<dim>>> cell_data_sorted;
+
+    unsigned int counter = 0;
+
+    for(const auto & cell : distributed_tria.cell_iterators_on_level(0))
+      cell_data_sorted.emplace_back(cell->id().get_coarse_cell_id(), cell_data[counter++]);
+
+    std::sort(cell_data_sorted.begin(), cell_data_sorted.end(), [](const auto & a, const auto & b) {
+      return a.first < b.first;
+    });
+
+    cell_data.clear();
+
+    for(const auto & i : cell_data_sorted)
+      cell_data.emplace_back(i.second);
+
+    serial_tria.create_triangulation(points, cell_data, sub_cell_data);
+  }
+
+  // execute refinement on first process of node if there is refinement left
+  if(distributed_tria.n_global_levels() > 1)
+  {
+    // collect refinement flags from distributed triangulation first on rank 0
+    std::vector<std::vector<std::vector<CellId>>> refinement_flags(
```

Comment:
I wonder why we have three nested vectors, below I only see two???

Reply:
The vector you probably do not see is what gets created within `Utilities::MPI::gather`, a vector for the vector collected on each MPI process. So the three vectors are level, the MPI process origin, and the cells that should be refined. There is certainly redundant information in the second and third vectors, as the ghosts and owned cells are seen by more than one rank, but that would only increase the code size.